### ISSUE TITLE
Tor-only transaction broadcast onlynet=onion alternative

### ISSUE THREAD

**Issue by jb55** [bitcoin/bitcoin#19042]
It would be nice if there was a way to tell core to only broadcast transactions over Tor, but use clearnet for everything else.

As far as I can tell the only way to guarantee transaction broadcast over Tor is to use onlynet=onion, which means that all network traffic goes over Tor. The problem is that this introduces latency for blocks/confirmations and transactions. Some users might only want to use Tor to try to protect "first broadcast" transaction source analysis which can be a problem with sybil attacking clearnet nodes. This can still be a problem on Tor as well, but can be more difficult to pull off.

Some bitcoin software such as wasabi and joinmarket already implement this by establishing a new Tor circuit, broadcasting the transaction, and then disconnecting. Perhaps bitcoin could do this as well?

Some things that were brought up when I mentioned this on IRC by @naumenkogs, @luke-jr  and others:

For the resistance to first-broadcast analysis to be effective, the node should treat the transaction as if it wasn't theirs to begin with. I'm not familiar with the network code, so I'm not sure if anything would have to change relay policy-wise.

Would be happy to hear other thoughts on what would be required to implement this, or if there's something I'm missing, as it's something I would be interested in implementing if there are enough concept acks.

TL;DR rationale:

- A nodes wants to recieve blocks/confirmations/transactions as soon as possible with the lowest latency

- A node wants increased broadcast privacy without paying the full cost of onlynet=tor latency

**Comment by vedran6** [comment]
`onlynet=onion` guarantees that outgoing connections are made through Tor; incoming clearnet connections are still allowed. I think there's no reason to exclude non-broadcast activity from the Tor segment since being connected to more networks can only increase partitioning resistance.

That said, I think that `onlynet=onion` can create false expectations in terms of privacy because a node operating on both networks (due to incoming connections) will still broadcast through the clearnet.

It would be good to have something like: `broadcast-network=ipv4,ipv6,onion` and make it inherit the `onlynet` parameter, or else default to Tor if Tor is running locally.

Would there be any interest in this?

**Comment by kristapsk** [comment]
> Some bitcoin software such as wasabi and joinmarket already implement this by establishing a new Tor circuit, broadcasting the transaction, and then disconnecting.

No, JoinMarket does not do that. Broadcast happens through Bitcoin Core. But not always by the initiator of the coinjoin (taker), by default it will be one of the participiants in coinjoin randomly.

**Comment by MarcoFalke** [comment]
> For the resistance to first-broadcast analysis to be effective, the node should treat the transaction as if it wasn't theirs to begin with. I'm not familiar with the network code, so I'm not sure if anything would have to change relay policy-wise.

I think the general idea is that you should consider your mempool to be an open book to all your inbound and outbound peers. So adding a wallet transaction to the mempool, which is readable by all your clearnet peers, and then sending the transaction on tor only will make it more than trivial for all clearnet peers to mark all your wallet transactions. See https://github.com/bitcoin/bitcoin/pull/15668#issuecomment-476738512 .

So I think one solution would be to never put the transaction in the mempool, but have logic in the wallet to create a one-shot connection to spit out the tx and immediately disconnect. Obviously your peer will learn your network fingerprint (version message), as well as your wallet tx (tx message). Though, with a connection over tor, and a normalized version message this should be conceptually OK.

The reason why I haven't drafted an implementation yet is because the wallet can't keep track of non-mempool change. See #11887. And fixing that appears non-trivial at first glance.

**Comment by jb55** [comment]
> > Some bitcoin software such as wasabi and joinmarket already implement this by establishing a new Tor circuit, broadcasting the transaction, and then disconnecting.
> 
> No, JoinMarket does not do that. Transaction broadcast happens through Bitcoin Core. But not always by the initiator of the coinjoin (taker), by default it will be one of the participiants in coinjoin randomly.

interesting, when I wrote this I distinctly remember someone telling me that's how it worked and showed me the code that does it... perhaps I misunderstood the details. thanks for the clarification.

**Comment by kristapsk** [comment]
> I distinctly remember someone telling me that's how it worked and showed me the code that does it...

Wasabi does that, there you are correct, but not JoinMarket.

**Comment by kristapsk** [comment]
With default config in JoinMarket tx broadcast also is done by random peer, we just don't do multiple Tor circuit thing. Communication anyway happens not directly between peers, but via IRC server.

**Comment by prayank23** [comment]
> Communication anyway happens not directly between peers, but via IRC server.

Is this being replaced with lightning?

**Comment by kristapsk** [comment]
> Is this being replaced with lightning?

[Work in progress](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1000). But not replaced, both can exist in parallel.

**Comment by vasild** [comment]
Implemented in https://github.com/bitcoin/bitcoin/pull/29415 for the `sendrawtransaction` RPC. Not for the wallet yet. For the wallet, one could create unsigned PSBT from the GUI and then use the `walletprocesspsbt` and `sendrawtransaction` RPCs.

**Comment by Nonthawat100iQ** [comment]
NONTHAWAT PHUPHASIT

**Comment by jb55** [comment]
amazing, thanks guys!
